utils-zookeeper/src/main/java/cz/o2/proxima/utils/zookeeper/ZKGlobalWatermarkTracker.java

```diff
@@ -0,0 +1,458 @@
+    client = null;
+    retry.run();
+  }
+
+  private void handleBadVersion(String path, CompletableFuture<Void> res, Runnable onSuccess) {
+    getNodeVersion(path)
+        .whenComplete(
+            (v, exc) -> {
+              if (exc != null) {
+                res.completeExceptionally(exc);
+              } else {
+                updateVersion = v;
+                onSuccess.run();
+              }
+            });
+  }
+
+  private synchronized void createParentIfNotExists(String node) throws InterruptedException {
+    try {
+      if (!parentCreated) {
+        client().create(node, new byte[] {}, Ids.OPEN_ACL_UNSAFE, CreateMode.CONTAINER);
+        createWatchForParentNode(node);
+        parentCreated = true;
+      }
+    } catch (KeeperException ex) {
+      if (ex.code() == Code.CONNECTIONLOSS || ex.code() == Code.SESSIONEXPIRED) {
+        client = null;
+        createParentIfNotExists(node);
+      } else if (ex.code() != Code.NODEEXISTS) {
+        throw new RuntimeException(ex);
+      } else {
+        createWatchForParentNode(node);
+      }
+    }
+  }
+
+  private static byte[] longAsBytes(long watermark) {
+    return String.valueOf(watermark).getBytes(StandardCharsets.UTF_8);
+  }
+
+  private static long longFromBytes(byte[] bytes) {
+    return Long.parseLong(new String(bytes, StandardCharsets.UTF_8));
+  }
+
+  private String getParentNode() {
+    return parentNode + this.trackerName;
+  }
+
+  private synchronized void updatePartialWatermark(String name, long watermark) {
+    partialWatermarks.compute(
+        name, (k, value) -> Math.max(value == null ? Long.MIN_VALUE : value, watermark));
+
+    globalWatermark.set(
+        partialWatermarks.values().stream().min(Long::compare).orElse(Long.MIN_VALUE));
+
+    Optional.ofNullable(updatedFutures.remove(name))
+        .ifPresent(l -> l.forEach(f -> f.complete(null)));
+  }
+
+  private ZooKeeper client() {
+    if (client == null) {
+      synchronized (this) {
+        if (client == null) {
+          client = createNewZooKeeper();
+        }
+      }
+    }
+    return client;
+  }
+
+  @VisibleForTesting
+  ZooKeeper createNewZooKeeper() {
+    CountDownLatch connectLatch = new CountDownLatch(1);
+    ZooKeeper zoo =
+        ExceptionUtils.uncheckedFactory(
+            () ->
+                new ZooKeeper(
+                    Objects.requireNonNull(zkConnectString),
+                    sessionTimeout,
+                    getWatcher(connectLatch)));
+    ExceptionUtils.ignoringInterrupted(
+        () -> {
+          if (!connectLatch.await(10, TimeUnit.SECONDS)) {
+            throw new RuntimeException("Timeout while connecting to ZK");
+          }
+        });
+    return zoo;
+  }
+
+  @VisibleForTesting
+  Watcher getWatcher(CountDownLatch connectLatch) {
+    return event -> waitTillConnected(event, connectLatch::countDown);
+  }
+
+  private static void waitTillConnected(WatchedEvent event, Runnable whenConnected) {
```

Comment:
can be merged with getWatcher method

Reply:
Yes, it can, but I prefer this for readability purposes.